### PR TITLE
RBAC for admin apis

### PR DESCRIPTION
**Pull request by viv-defy**

### CODE REVIEW

File: bundler/src/handlers/admin.rs [hunk 1]
```diff
@@ -19,7 +18,7 @@ pub async fn topup_paymaster_deposit(
     req: HttpRequest,
     paymaster: Path<String>,
 ) -> Result<Json<BaseResponse<TransferResponse>>, ApiError> {
-    if Constants::ADMIN != get_user(req) {
+    if !is_admin(get_user(req)) {
```
Comment (ska1296) [inline bundler/src/handlers/admin.rs:21]: if you always check for "not" condition, I think it's better to change your function name and return value

File: bundler/src/handlers/admin.rs [hunk 2]
```diff
@@ -49,9 +48,15 @@ pub async fn add_currency_metadata(
...
         return Err(ApiError::BadRequest("Invalid credentials".to_string()));
     }
     let response = service.add_currency_metadata(body.into_inner()).await?;
     respond_json(response)
 }
+
+fn is_admin(user: String) -> bool {
+    let admins_env = std::env::var("ADMIN").expect("ADMIN env variable not set");
+    let admins = admins_env.split(",").collect::<Vec<&str>>();
+    admins.contains(&&*user)
```
Comment (ska1296) [inline bundler/src/handlers/admin.rs:61]: you can just do a string contains here. No need to convert it to an array.

**Review by ska1296 — CHANGES_REQUESTED** [review]

File: bundler/src/handlers/admin.rs (hunk 2 above)
Comment (ska1296) [inline bundler/src/handlers/admin.rs]:
```suggestion
    let admins = admins_env.split(",").collect::<Vec<String>>();
    admins.contains(&user)
```

File: bundler/src/handlers/admin.rs [hunk 3]
```diff
@@ -49,9 +48,15 @@ pub async fn add_currency_metadata(
...
-    if Constants::ADMIN != get_user(req) {
+    if is_not_admin(get_user(req)) {
         return Err(ApiError::BadRequest("Invalid credentials".to_string()));
     }
     let response = service.add_currency_metadata(body.into_inner()).await?;
     respond_json(response)
 }
+
+fn is_not_admin(user: String) -> bool {
+    let admins_env = std::env::var("ADMIN").expect("ADMIN env variable not set");
```
Comment (vishnukvnvv) [inline bundler/src/handlers/admin.rs:59]: Can be fetched once at server start and reused

**Comment by vishnukvnvv** [comment]
@viv-defy, Update your local branch from the release branch

File: bundler/src/services/admin_service.rs [hunk 4]
```diff
@@ -22,6 +22,7 @@ use crate::CONFIG;
 
 #[derive(Clone)]
 pub struct AdminService {
+    pub admins: Vec<String>,
```
Comment (vishnukvnvv) [inline bundler/src/services/admin_service.rs:25]: Remove this from the service layer and add it in config/settings

File: bundler/src/server.rs [hunk 5]
```diff
@@ -103,6 +103,13 @@ pub async fn init_services() -> ToadService {
         abi: simple_account_factory.clone(),
     };
 
+    // admins
+    let admins_env = std::env::var("ADMIN").expect("ADMIN env variable not set");
+    let admins = admins_env
+        .split(",")
+        .map(|s| s.to_string())
+        .collect::<Vec<String>>();
+
```
Comment (ska1296) [inline bundler/src/server.rs:112]: move admin as a variable to "settings" struct instead of passing it around to services

**Review by ska1296 — APPROVED** [review]

**Review by vishnukvnvv — APPROVED** [review]

**Pull request merged**